README.md
Working on an app of mine, I needed a way to display images in a masonry style. Additionally I needed
the display to change based on how big the user wants the thumbnails. I tried to make it as close 
as possible to Samsung's gallery display, since I find intuitive.

## Features

- Masonry style display when imagesPerRow is set to 2
- Can go up to 8 images per row, images will be cropped and resized accordingly.
- Can go down to 1 image, images will be displayed based on the width of the screen.
- Can load local images, by setting pathOrUrl to a local path, and setting localOrRemote to 'local'
- Can load network images, by setting pathOrUrl to a remote url, and setting localOrRemote to 'remote'

## Usage

```dart
// Define a Map with String image name and List<double> of image width x height
Map<String, List<double>> images = {
  'feed8.jpg': [4032, 3024],
  'feed1.jpg': [3024, 4032],
  'feed2.jpg': [3024, 4032],
  'feed3.jpg': [3024, 3024],
  'feed4.jpg': [4032, 1816],
  'feed5.jpg': [3024, 3024],
  'feed6.jpg': [3024, 3024],
  'feed7.jpg': [3024, 3024],
};

// Include this in the child of something else.
Gallery(
buildContext: context,
imageWithSizesMap: images,
totalSidesPadding: 48,// Sum of the left and right padding if any, defaults to 0
callBack: callbackFunctionName //callback function will be called with a String of the path of the clicked image
)
.galleryImages(
imagesPerRow: 2, pathOrUrl: 'lib/assets/images/', localOrRemote: 'local'),

```



CHANGELOG.md
## 0.0.91

* if image name starts with http:// or https:// always load a network image.

## 0.0.9

* add the ability to add a widget above the image.


## 0.0.8

* added base64 support.


## 0.0.7

* added the ability to set a callback function on images tap, when an image is tapped, the 
* callback function will be called with the path of the tapped image as a string.

## 0.0.6

* added the ability to have external padding. you can define totalSidePadding as the total of left
* and right padding.

## 0.0.5

* remove debug print

## 0.0.4

* bug fix

## 0.0.3

* just a bit of refactoring. (DON'T USE, it is bugged.)

## 0.0.2

* just a bit of refactoring.

## 0.0.1

* A small neat package to arrange gallery images based on their size and how many images per row.


test/gallery_images_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:gallery_images/gallery_images.dart';

void main() {
  // test('adds one to input values', () {
  //   final calculator =
  //       Gallery(buildContext: context, imageWithSizesMap: imageWithSizesMap);
  //   expect(calculator.addOne(2), 3);
  //   expect(calculator.addOne(-7), -6);
  //   expect(calculator.addOne(0), 1);
  // });
}


test/images_gallery_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:images_gallery/images_gallery.dart';

void main() {
  // test('adds one to input values', () {
  //   final calculator =
  //       Gallery(buildContext: context, imageWithSizesMap: imageWithSizesMap);
  //   expect(calculator.addOne(2), 3);
  //   expect(calculator.addOne(-7), -6);
  //   expect(calculator.addOne(0), 1);
  // });
}


lib/gallery_images.dart
library gallery_images;

import 'package:flutter/material.dart';

/// A Gallery Display.
class Gallery {
  late final BuildContext context;
  late final Map<String, List<double>> images;
  Gallery(
      {required BuildContext buildContext,
      required Map<String, List<double>> imageWithSizesMap}) {
    context = buildContext;
    images = imageWithSizesMap;
  }
  Widget normalGalleryImages(
      {required Map<String, List<double>> images,
      required String pathOrUrl,
      String localOrRemote = 'local',
      int imagesPerRow = 2,
      double padding = 3}) {
    List<Widget> gallery = [];
    List<Widget> imageWidgets = [];
    double screenWithWithoutPadding =
        MediaQuery.of(context).size.width - (imagesPerRow * 2 * padding);
    double width = (screenWithWithoutPadding) / imagesPerRow;
    double height = width;
    List<String> imageNames = images.keys.toList();
    for (int i = 0; i < imageNames.length; i++) {
      var thisImageSize = images[imageNames[i]];
      print((thisImageSize![0] / screenWithWithoutPadding));
      imageWidgets.add(
        Center(
          child: Padding(
            padding: EdgeInsets.all(padding),
            child: Container(
              height: imagesPerRow > 1
                  ? height
                  : thisImageSize![1] /
                      (thisImageSize![0] / screenWithWithoutPadding),
              width: width,
              clipBehavior: Clip.hardEdge,
              decoration: BoxDecoration(
                image: DecorationImage(
                  fit: isVertical(thisImageSize!) || imagesPerRow == 1
                      ? BoxFit.fitWidth
                      : BoxFit.fitHeight,
                  image: localOrRemote == 'remote'
                      ? NetworkImage('$pathOrUrl${imageNames[i]}')
                      : AssetImage('$pathOrUrl${imageNames[i]}')
                          as ImageProvider,
                ),
              ),
            ),
          ),
        ),
      );
      if ((i + 1) % imagesPerRow == 0) {
        gallery.add(Row(
          children: imageWidgets,
        ));
        imageWidgets = [];
      }
    }
    if (imageWidgets.isNotEmpty) {
      gallery.add(Row(
        children: imageWidgets,
      ));
    }
    return SingleChildScrollView(
      scrollDirection: Axis.vertical,
      child: Column(
        children: gallery,
      ),
    );
  }

  Widget galleryImages(
      {required String pathOrUrl,
      String localOrRemote = 'local',
      int imagesPerRow = 2,
      double padding = 4}) {
    localOrRemote = localOrRemote.toLowerCase();
    pathOrUrl = fixPathOrUrl(pathOrUrl);
    if (imagesPerRow != 2) {
      return normalGalleryImages(
          images: images,
          imagesPerRow: imagesPerRow,
          padding: padding,
          pathOrUrl: pathOrUrl,
          localOrRemote: localOrRemote);
    }
    List<Widget> gallery = [];
    List<String> imageNames = images.keys.toList();

    localOrRemote = localOrRemote.toLowerCase();
    double totalHeight = MediaQuery.of(context).size.height;
    double calcTotalHeight = 0;
    int lastI = 0;
    double top = 0;
    double left = 0;
    double imagesHeight = 0;

    for (int i = 0; i < imageNames.length; i++) {
      if (i < imageNames.length - 1) {
        String currentImage = imageNames[i];
        String nextImage = imageNames[i + 1];
        List<double> positions =
            getPositioning(images[currentImage]!, images[nextImage]!);

        if (positions.length == 2) {
          imagesHeight = positions[1] /
              (images[currentImage]![0] / MediaQuery.of(context).size.width);
          gallery.add(
            Positioned(
              top: top,
              left: 0,
              width: MediaQuery.of(context).size.width,
              height: imagesHeight,
              child: Padding(
                padding: EdgeInsets.all(padding),
                child: localOrRemote == 'remote'
                    ? Image.network(
                        '$pathOrUrl$currentImage',
                        fit: BoxFit.fill,
                      )
                    : Image.asset(
                        '$pathOrUrl$currentImage',
                        fit: BoxFit.fill,
                      ),
              ),
            ),
          );
          top += imagesHeight;
        } else {
          gallery.add(Positioned(
            top: top,
            left: 0,
            width: MediaQuery.of(context).size.width * positions[0],
            height: positions[2],
            child: Padding(
              padding: EdgeInsets.all(padding),
              child: localOrRemote == 'remote'
                  ? Image.network(
                      '$pathOrUrl$currentImage',
                      fit: BoxFit.fill,
                    )
                  : Image.asset(
                      '$pathOrUrl$currentImage',
                      fit: BoxFit.fill,
                    ),
            ),
          ));
          left = MediaQuery.of(context).size.width * positions[0];
          gallery.add(
            Positioned(
              top: top,
              left: left,
              width: MediaQuery.of(context).size.width * positions[1],
              height: positions[2],
              child: Padding(
                padding: EdgeInsets.all(padding),
                child: localOrRemote == 'remote'
                    ? Image.network(
                        '$pathOrUrl$currentImage',
                        fit: BoxFit.fill,
                      )
                    : Image.asset(
                        '$pathOrUrl$currentImage',
                        fit: BoxFit.fill,
                      ),
              ),
            ),
          );
          left = 0;
          top += positions[2];
          calcTotalHeight += top;
          i++;
        }
      }
      lastI = i;
    }
    if (lastI == images.length - 1) {
      imagesHeight = images[imageNames.last]![1] /
          (images[imageNames.last]![0] / (MediaQuery.of(context).size.width));
      gallery.add(
        Positioned(
          width: MediaQuery.of(context).size.width,
          height: imagesHeight,
          top: top,
          left: 0,
          child: localOrRemote == 'remote'
              ? Image.network(
                  '$pathOrUrl${imageNames.last}',
                  fit: BoxFit.fill,
                )
              : Image.asset(
                  '$pathOrUrl${imageNames.last}',
                  fit: BoxFit.fill,
                ),
        ),
      );
      calcTotalHeight = top + imagesHeight;
    }
    gallery.add(Container(
      height: calcTotalHeight > totalHeight ? calcTotalHeight : totalHeight,
    ));

    return SingleChildScrollView(
      scrollDirection: Axis.vertical,
      child: Stack(
        children: gallery,
      ),
    );
  }

  List<double> getPositioning(List<double> size1, List<double> size2) {
    // sizes will contain the width of the 1st image, width of 2nd image, and
    // height of both images

    List<double> sizes = [1, size1[1]];

    if (isSquare(size1) && isSquare(size2) ||
        (isHorizontal(size1) && isHorizontal(size2)) ||
        (isVertical(size1) && isVertical(size2))) {
      double firstHeight =
          size1[1] / (size1[0] / (MediaQuery.of(context).size.width / 2));
      double secondHeight =
          size2[1] / (size2[0] / (MediaQuery.of(context).size.width / 2));

      return [
        0.5,
        0.5,
        firstHeight > secondHeight ? secondHeight : firstHeight
      ];
    }

    if ((isSquare(size1) || isHorizontal(size1)) && isVertical(size2)) {
      double firstHeight =
          size1[1] / (size1[0] / (MediaQuery.of(context).size.width * 0.62));
      double secondHeight =
          size2[1] / (size2[0] / (MediaQuery.of(context).size.width * 0.38));
      return [
        0.62,
        0.38,
        firstHeight > secondHeight ? secondHeight : firstHeight
      ];
    }

    if ((isSquare(size2) || isHorizontal(size2)) && isVertical(size1)) {
      double firstHeight =
          size1[1] / (size1[0] / (MediaQuery.of(context).size.width * 0.38));
      double secondHeight =
          size2[1] / (size2[0] / (MediaQuery.of(context).size.width / 0.62));
      return [
        0.38,
        0.62,
        firstHeight > secondHeight ? secondHeight : firstHeight
      ];
    }

    return sizes;
  }

  String fixPathOrUrl(String pathOrUrl) {
    if (pathOrUrl.substring(pathOrUrl.length - 1) != '/') {
      pathOrUrl += '/';
    }
    return pathOrUrl;
  }

  bool isSquare(List<double> size) => size[0] == size[1];

  bool isHorizontal(List<double> size) => size[0] > size[1];

  bool isVertical(List<double> size) => size[0] < size[1];
}


lib/images_gallery.dart
library images_gallery;

import 'dart:convert';

import 'package:flutter/material.dart';

/// A Gallery Display.
class Gallery {
  late final BuildContext context;
  late final Map<String, List<double>> images;
  late final Map<String, String> imagesBase64;
  late final Map<String, Widget> imagesChildren;
  late final Map<String, Alignment> childrenAlignment;
  late final double remainingScreenWidth;
  late final Function? callBack;
  Gallery(
      {required BuildContext buildContext,
      required Map<String, List<double>> imageWithSizesMap,
      double totalSidesPadding = 0,
      this.imagesBase64 = const {},
      this.imagesChildren = const {},
      this.childrenAlignment = const {},
      this.callBack}) {
    context = buildContext;
    images = imageWithSizesMap;
    remainingScreenWidth =
        MediaQuery.of(context).size.width - totalSidesPadding;
  }
  // This function is called if imagesPerRow isn't 2
  Widget _normalGalleryImages(
      {required Map<String, List<double>> images,
      required String pathOrUrl,
      String localOrRemote = 'local',
      int imagesPerRow = 3,
      double padding = 3}) {
    List<Widget> gallery = [];
    List<Widget> imageWidgets = [];
    // Get screen width and subtract the padding, to calculate image width
    double screenWithWithoutPadding =
        remainingScreenWidth - (imagesPerRow * 2 * padding);
    double width = (screenWithWithoutPadding) / imagesPerRow;
    double height = width;
    List<String> imageNames = images.keys.toList();
    for (int i = 0; i < imageNames.length; i++) {
      var thisImageSize = images[imageNames[i]];
      var imageWidget = Center(
        child: Padding(
          padding: EdgeInsets.all(padding),
          child: Container(
            alignment: childrenAlignment.keys.contains(imageNames[i])
                ? childrenAlignment[imageNames[i]]
                : Alignment.bottomCenter,
            height: imagesPerRow > 1
                ? height
                : thisImageSize![1] /
                    (thisImageSize![0] / screenWithWithoutPadding),
            width: width,
            clipBehavior: Clip.hardEdge,
            decoration: BoxDecoration(
              image: DecorationImage(
                fit: isVertical(thisImageSize!) || imagesPerRow == 1
                    ? BoxFit.fitWidth
                    : BoxFit.fitHeight,
                image: imagesBase64.keys.contains(imageNames[i])
                    ? MemoryImage(base64Decode(imagesBase64[imageNames[i]]!))
                    : localOrRemote == 'remote' ||
                            imageNames[i].startsWith("http://") ||
                            imageNames[i].startsWith("https://")
                        ? NetworkImage('$pathOrUrl${imageNames[i]}')
                        : AssetImage('$pathOrUrl${imageNames[i]}')
                            as ImageProvider,
              ),
            ),
            child: imagesChildren.keys.contains(imageNames[i])
                ? imagesChildren[imageNames[i]]
                : Container(),
          ),
        ),
      );
      if (callBack != null) {
        imageWidgets.add(GestureDetector(
          onTap: () {
            if (imagesBase64.keys.contains(imageNames[i])) {
              callBack!(imagesBase64[imageNames[i]]);
            } else {
              callBack!("$pathOrUrl${imageNames[i]}");
            }
          },
          child: imageWidget,
        ));
      } else {
        imageWidgets.add(imageWidget);
      }

      if ((i + 1) % imagesPerRow == 0) {
        gallery.add(Row(
          children: imageWidgets,
        ));
        imageWidgets = [];
      }
    }

    if (imageWidgets.isNotEmpty) {
      gallery.add(Row(
        children: imageWidgets,
      ));
    }
    return SingleChildScrollView(
      scrollDirection: Axis.vertical,
      child: Column(
        children: gallery,
      ),
    );
  }

  // This is the main function that will be called
  Widget galleryImages({
    required String pathOrUrl,
    String localOrRemote = 'local',
    int imagesPerRow = 2,
    double padding = 4,
  }) {
    localOrRemote = localOrRemote.toLowerCase();
    pathOrUrl = fixPathOrUrl(pathOrUrl);
    if (imagesPerRow != 2) {
      return _normalGalleryImages(
          images: images,
          imagesPerRow: imagesPerRow,
          padding: padding,
          pathOrUrl: pathOrUrl,
          localOrRemote: localOrRemote);
    }

    List<Widget> gallery = [];
    List<String> imageNames = images.keys.toList();

    localOrRemote = localOrRemote.toLowerCase();
    double totalHeight = MediaQuery.of(context).size.height;
    double calcTotalHeight = 0;
    int lastI = 0;
    double top = 0;
    double left = 0;
    double imagesHeight = 0;

    for (int i = 0; i < imageNames.length; i++) {
      if (i < imageNames.length - 1) {
        String currentImage = imageNames[i];
        String nextImage = imageNames[i + 1];
        List<double> positions =
            getPositioning(images[currentImage]!, images[nextImage]!);

        if (positions.length == 2) {
          imagesHeight =
              positions[1] / (images[currentImage]![0] / remainingScreenWidth);
          var imageWidget = Container(
            alignment: childrenAlignment.keys.contains(currentImage)
                ? childrenAlignment[currentImage]
                : Alignment.bottomCenter,
            margin: EdgeInsets.all(padding),
            decoration: BoxDecoration(
              image: DecorationImage(
                image: imagesBase64.keys.contains(currentImage)
                    ? MemoryImage(base64Decode(imagesBase64[currentImage]!))
                    : localOrRemote == 'remote' ||
                            currentImage.startsWith("http://") ||
                            currentImage.startsWith("https://")
                        ? NetworkImage('$pathOrUrl$currentImage')
                        : AssetImage('$pathOrUrl$currentImage')
                            as ImageProvider,
              ),
            ),
            child: imagesChildren.keys.contains(currentImage)
                ? imagesChildren[currentImage]
                : Container(),
          );
          if (callBack != null) {
            gallery.add(
              Positioned(
                top: top,
                left: 0,
                width: remainingScreenWidth,
                height: imagesHeight,
                child: GestureDetector(
                  onTap: () {
                    if (imagesBase64.keys.contains(currentImage)) {
                      callBack!(imagesBase64[currentImage]);
                    } else {
                      callBack!("$pathOrUrl$currentImage");
                    }
                  },
                  child: imageWidget,
                ),
              ),
            );
          } else {
            gallery.add(
              Positioned(
                  top: top,
                  left: 0,
                  width: remainingScreenWidth,
                  height: imagesHeight,
                  child: imageWidget),
            );
          }
          top += imagesHeight;
        } else {
          var imageWidget = Container(
            alignment: childrenAlignment.keys.contains(currentImage)
                ? childrenAlignment[currentImage]
                : Alignment.bottomCenter,
            margin: EdgeInsets.all(padding),
            decoration: BoxDecoration(
              image: DecorationImage(
                image: imagesBase64.keys.contains(currentImage)
                    ? MemoryImage(base64Decode(imagesBase64[currentImage]!))
                    : localOrRemote == 'remote' ||
                            currentImage.startsWith("http://") ||
                            currentImage.startsWith("https://")
                        ? NetworkImage('$pathOrUrl$currentImage')
                        : AssetImage('$pathOrUrl$currentImage')
                            as ImageProvider,
              ),
            ),
            child: imagesChildren.keys.contains(currentImage)
                ? imagesChildren[currentImage]
                : Container(),
          );
          if (callBack != null) {
            gallery.add(Positioned(
              top: top,
              left: 0,
              width: remainingScreenWidth * positions[0],
              height: positions[2],
              child: GestureDetector(
                onTap: () {
                  if (imagesBase64.keys.contains(currentImage)) {
                    callBack!(imagesBase64[currentImage]);
                  } else {
                    callBack!("$pathOrUrl$currentImage");
                  }
                },
                child: imageWidget,
              ),
            ));
          } else {
            gallery.add(
              Positioned(
                  top: top,
                  left: 0,
                  width: remainingScreenWidth * positions[0],
                  height: positions[2],
                  child: imageWidget),
            );
          }
          left = remainingScreenWidth * positions[0];
          imageWidget = Container(
            alignment: childrenAlignment.keys.contains(nextImage)
                ? childrenAlignment[nextImage]
                : Alignment.bottomCenter,
            margin: EdgeInsets.all(padding),
            decoration: BoxDecoration(
              image: DecorationImage(
                image: imagesBase64.keys.contains(nextImage)
                    ? MemoryImage(base64Decode(imagesBase64[nextImage]!))
                    : localOrRemote == 'remote' ||
                            nextImage.startsWith("http://") ||
                            nextImage.startsWith("https://")
                        ? NetworkImage('$pathOrUrl$nextImage')
                        : AssetImage('$pathOrUrl$nextImage') as ImageProvider,
              ),
            ),
            child: imagesChildren.keys.contains(nextImage)
                ? imagesChildren[nextImage]
                : Container(),
          );
          if (callBack != null) {
            gallery.add(Positioned(
              top: top,
              left: left,
              width: remainingScreenWidth * positions[1],
              height: positions[2],
              child: GestureDetector(
                onTap: () {
                  if (imagesBase64.keys.contains(nextImage)) {
                    callBack!(imagesBase64[nextImage]);
                  } else {
                    callBack!("$pathOrUrl$nextImage");
                  }
                },
                child: imageWidget,
              ),
            ));
          } else {
            gallery.add(
              Positioned(
                  top: top,
                  left: left,
                  width: remainingScreenWidth * positions[1],
                  height: positions[2],
                  child: imageWidget),
            );
          }
          left = 0;
          top += positions[2];
          calcTotalHeight += top;
          i++;
        }
      }
      lastI = i;
    }
    if (lastI == images.length - 1) {
      imagesHeight = images[imageNames.last]![1] /
          (images[imageNames.last]![0] / remainingScreenWidth);
      var imageWidget = Container(
        alignment: childrenAlignment.keys.contains(imageNames.last)
            ? childrenAlignment[imageNames.last]
            : Alignment.bottomCenter,
        margin: EdgeInsets.all(padding),
        decoration: BoxDecoration(
          image: DecorationImage(
            image: imagesBase64.keys.contains(imageNames.last)
                ? MemoryImage(base64Decode(imagesBase64[imageNames.last]!))
                : localOrRemote == 'remote' ||
                        imageNames.last.startsWith("http://") ||
                        imageNames.last.startsWith("https://")
                    ? NetworkImage('$pathOrUrl${imageNames.last}')
                    : AssetImage('$pathOrUrl${imageNames.last}')
                        as ImageProvider,
          ),
        ),
        child: imagesChildren.keys.contains(imageNames.last)
            ? imagesChildren[imageNames.last]
            : Container(),
      );
      if (callBack != null) {
        gallery.add(Positioned(
          width: remainingScreenWidth,
          height: imagesHeight,
          top: top,
          left: 0,
          child: GestureDetector(
            onTap: () {
              if (imagesBase64.keys.contains(imageNames.last)) {
                callBack!(imagesBase64[imageNames.last]);
              } else {
                callBack!("$pathOrUrl${imageNames.last}");
              }
            },
            child: imageWidget,
          ),
        ));
      } else {
        gallery.add(
          Positioned(
            width: remainingScreenWidth,
            height: imagesHeight,
            top: top,
            left: 0,
            child: imageWidget,
          ),
        );
      }
      calcTotalHeight = top + imagesHeight;
    }
    gallery.add(Container(
      height: calcTotalHeight > totalHeight ? calcTotalHeight : totalHeight,
    ));

    return SingleChildScrollView(
      scrollDirection: Axis.vertical,
      child: Stack(
        children: gallery,
      ),
    );
  }

  List<double> getPositioning(List<double> size1, List<double> size2) {
    // sizes will contain the width of the 1st image, width of 2nd image, and
    // height of both images

    List<double> sizes = [1, size1[1]];

    if (isSquare(size1) && isSquare(size2) ||
        (isHorizontal(size1) && isHorizontal(size2)) ||
        (isVertical(size1) && isVertical(size2))) {
      double firstHeight = size1[1] / (size1[0] / (remainingScreenWidth / 2));
      double secondHeight = size2[1] / (size2[0] / (remainingScreenWidth / 2));

      return [
        0.5,
        0.5,
        firstHeight > secondHeight ? secondHeight : firstHeight
      ];
    }

    if ((isSquare(size1) || isHorizontal(size1)) && isVertical(size2)) {
      double firstHeight =
          size1[1] / (size1[0] / (remainingScreenWidth * 0.62));
      double secondHeight =
          size2[1] / (size2[0] / (remainingScreenWidth * 0.38));
      return [
        0.62,
        0.38,
        firstHeight > secondHeight ? secondHeight : firstHeight
      ];
    }

    if ((isSquare(size2) || isHorizontal(size2)) && isVertical(size1)) {
      double firstHeight =
          size1[1] / (size1[0] / (remainingScreenWidth * 0.38));
      double secondHeight =
          size2[1] / (size2[0] / (remainingScreenWidth / 0.62));
      return [
        0.38,
        0.62,
        firstHeight > secondHeight ? secondHeight : firstHeight
      ];
    }

    return sizes;
  }

  String fixPathOrUrl(String pathOrUrl) {
    if (pathOrUrl.isNotEmpty &&
        pathOrUrl.substring(pathOrUrl.length - 1) != '/') {
      pathOrUrl += '/';
    }
    return pathOrUrl;
  }

  bool isSquare(List<double> size) => size[0] == size[1];

  bool isHorizontal(List<double> size) => size[0] > size[1];

  bool isVertical(List<double> size) => size[0] < size[1];
}


example/main.dart
import 'dart:core';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:images_gallery/images_gallery.dart';

class Feed extends StatefulWidget {
  const Feed({Key? key}) : super(key: key);

  @override
  State<Feed> createState() => _FeedState();
}

class _FeedState extends State<Feed> {
  int imagesPerRow = 2;
  double currentScale = 1;
  double updatedScale = 1;
  @override
  Widget build(BuildContext context) {
    Map<String, List<double>> images = {
      'feed8.jpg': [4032, 3024],
      'feed1.jpg': [3024, 4032],
      'feed2.jpg': [3024, 4032],
      'feed3.jpg': [3024, 3024],
      'feed4.jpg': [4032, 1816],
      'feed5.jpg': [3024, 3024],
      'feed6.jpg': [3024, 3024],
      'feed7.jpg': [3024, 3024],
    };
    // If some images are base64 you can send them in this dictionary where the key
    // is the same as images dictionary and the value is the base64 string
    Map<String, String> imagesBase64 = {'feed8.jpg': 'Some BASE64 string'};
    Map<String, Widget> imagesChildren = {
      'feed8.jpg': Container(
        child: Text('Play as you like'),
      )
    };
    Map<String, Alignment> childrenAlignment = {
      'feed8.jpg': Alignment.bottomCenter
    };

    void callbackFunctionName(String newPath) {
      //Do something with the imagePath newPath
    }

    return SafeArea(
      child: Gallery(
              buildContext: context,
              imageWithSizesMap: images,
              imagesBase64: imagesBase64, // send base64 images here (optional)
              imagesChildren:
                  imagesChildren, // if you need to add a layer above the image, a title or something (optional)
              childrenAlignment:
                  childrenAlignment, // if children are provided, alignment can be controlled, default Alignment.bottomCenter (optional)
              totalSidesPadding:
                  48, // Sum of the left and right padding if any, defaults to 0
              callBack:
                  //callback function will be called with a String of the path of the clicked image,
                  // if the image is base64, the base 64 string will be sent to the callback
                  callbackFunctionName)
          .galleryImages(
              imagesPerRow: imagesPerRow,
              pathOrUrl: 'lib/assets/images/',
              localOrRemote: 'local'),
    );
  }
}
